File: DevDayAI2019/Neurona_Unica.m
%Carga de datos de entrenamiento
load('desgrad_dataset.mat')

%Neurona artificial
% x[0]--w[1]--
%              \
%                + --f(net)-- y
%              /
% x[1]--w[1]--

%Inicializacion de pesos (aleatorios)
xmin=-1
xmax=1
w=xmin+rand(1,2)*(xmax-xmin) %vector de pesos dimencion [1,2]
b = rand() % bias


%% Entrenamiento
t = dataset(:,3);  %target
p = dataset(:,1:2); %patrones

N = 100; %Numero de epocas de entrenamiento
a = 0.1; %Factor de aprendisaje
for Ep = 1:N
    for k = 1:numel(t)
        net = dot(w,p(k,:))+b;
        y = logsig(net);
        df = logsig(net)*(1-logsig(net));
        e = t(k)- y;
        w = w + a*e*df*p(k,:);
        b = b + a*e*df;
    end
end

%% Graficas
%Mapa de prediccion
x0 = linspace(-1,7);
x1 = linspace(-1,7);
[X0,X1] = meshgrid(x0,x1);

for i = 1:size(X0,1)
    for j = 1:size(X0,2)
        Y(i,j) = logsig(dot(w,[X0(i,j),X1(i,j)])+b);
    end
end

contourf(X0,X1,Y,10)
map =[0 0 1
      1 0 0];
colormap(map)
%%
hold on
for i = 1:size(dataset,1)
    if(dataset(i,3)==1)
        scatter(dataset(i,1),dataset(i,2),'MarkerEdgeColor','k','MarkerFaceColor','r')      
    else
        scatter(dataset(i,1),dataset(i,2),'MarkerEdgeColor','k','MarkerFaceColor','b')
    end
end
hold off
xlabel('x')
ylabel('y')
grid on 

File: DevDayAI2019/Neurona_Unica.py
# -*- coding: utf-8 -*-
"""
Created on Mon Nov 11 20:48:37 2019

@author: Rodolfo Escobar
"""

import numpy as np
import matplotlib.pyplot as plt

def logsig(x):
    return 1/(1+np.exp(-x))

def df(z):
    a = logsig(z)
    return a * (1.0 - a)

dataset = np.load("desgrad_dataset.npy")

p = dataset[:,0:2]
t = dataset[:,2]

def DG(p,t,Ep=10):
    """
    Entrenamiento por Decenso de Gradiente
    """
    m,n = p.shape
    a = 0.5
    
    #--- Pesos iniciales ---
    w = np.random.uniform(-0.25,0.25,2)
    b = np.random.uniform(-0.25,0.25)
    # ----------------------
    
    Delta = np.zeros([m,2])

    for N in range(Ep):
        for ti in range(m):
            #--- Salida
            net = np.dot(w,p[ti])+b
            y = logsig(net)
            #----

            err = t[ti]- y             
            Delta = 2*err*df(net)*p[ti]
            w = w + a*Delta
            b = b + a*2*err*df(net)
              
        w += a*Delta.mean(axis=0)
        b += a*Delta.mean(axis=0)
        
    return w,b

def modelo(X,w,b):
    z = np.dot(w,X)+b
    return logsig(z)
    
#--- Entrenamiento --- 
w,b = DG(p,t,100)

#%%---- Graficas ---
def hardlim(z,lim = 0):
    return 1 if z>lim else 0
#------------
NN = 20
x0 = np.linspace(p.min()-0.5,p.max()+0.5,NN)
x1 = np.linspace(p.min()-0.5,p.max()+0.5,NN)
Zp = np.zeros([NN,NN])

plt.figure(1)
for i in range(NN):
    for j in range(NN):             #-----x------#
        Zp[i][j] = hardlim(np.dot(w,[x0[j],x1[i]])+b)#hardlim(modelo([Xt[j],Yt[i]],w,b),0.5)

X, Y = np.meshgrid(x0, x1)        
plt.contourf(X,Y,Zp,cmap = 'bwr',alpha=0.6)
plt.grid(True)
plt.xlabel("$x_0$")
plt.ylabel("$x_1$")        
#----------------------

for i in range(len(t)):
    if(t[i]==1):
        plt.scatter(p[i,0],p[i,1],c='r')
    else:
        plt.scatter(p[i,0],p[i,1],c='b')